packages/alfa-parser/src/parser.ts

```diff
@@ -149,4 +149,42 @@ export namespace Parser {
       flatMap(separator, () => map(right, right => [left, right]))
     );
   }
+
+  export function separatedList<I, T, E>(
+    parser: Parser<I, T, E>,
+    separator: Parser<I, unknown, E>
+  ): Parser<I, Iterable<T>, E> {
+    return input => {
+      const values: Array<T> = [];
+      const result = parser(input);
+
+      if (result.isErr()) {
+        return result;
+      }
+
+      let value: T;
+
+      [input, value] = result.get();
+      values.push(value);
+
+      while (true) {
+        const separated = separator(input);
+
+        if (separated.isErr()) {
+          return Ok.of([input, values] as const);
+        }
+
+        [input] = separated.get();
+
+        const result = parser(input);
+
+        if (result.isErr()) {
+          return result;
+        }
+
+        [input, value] = result.get();
+        values.push(value);
+      }
```

Comment:
Consider the simplest case: `a`. When the separator fails, the parsing ends. In that case, the separator will fail and the parsing will end even though it's a valid list, albeit with a single element. Either way, I missed the obvious answer: We already have combinators for expressing this particular combinator so there's no need to spell it out 😄 It's just a matter of doing `pair(parser, zeroOrMore(right(separator, parser)))`.

Reply:
And this case would fail for the code in https://github.com/Siteimprove/alfa/pull/186#discussion_r377591655: `a,`. This would eat the trailing `,` which is not what we want.